### DISCUSSION TITLE
Threads executing event loop just like in the Qt framework

### DISCUSSION THREAD

**Discussion by embeddedmz** [dotnet/runtime#60474] [General]
Hi,

With MS technologies, there's only the GUI thread that has this capability. In other words, I want to have a thread that runs an "event loop" (or a pump message in other terms) as its main function and offer to the user methods like Invoke and BeginInvoke (like the methods that can be called on a WPF dispatcher).

I know that it can be made with a BlockingCollection<Action> and other threads can use a lambda to package a method and its arguments but it would be cool to have something operational right out of the box (and that we can use to get back the result, the return of the method, from the receiver thread). It reminds me of Unix processes that communicate with each other via message queues, I think this is safer than a program containing a spaghetti of threads using locks (managed by the user), I think that the first case is more structured and robust.

Some resources :
https://doc.qt.io/qt-5/threads-qobject.html
https://stackoverflow.com/questions/67418826/possibility-of-converting-a-c-qt-threads-using-event-loops-to-c-sharp-or-java
https://stackoverflow.com/questions/41271071/how-to-queue-up-delegates-to-be-executed-in-series-in-the-background-with-c

Best regards.

**Comment by HaloFour** [comment]
Are you looking for some kind of language feature for this?  If so, what would something like that look like?  This seems like more a question for the [Runtime repository](https://github.com/dotnet/runtime) that deals with the base APIs.

**Reply by embeddedmz** [reply]
Maybe both, I don't know which is the best candidate.

In Qt, a signal (i.e. an event) can be connected to a slot (delegate) with a queued or direct connection (queued : the event loop of a thread will execute it when it's possible, direct : the delegate will be executed in the same thread where the event is fired).

In C# adding a delegate to an event is agnostic of the the thread affinity that's why people use this trick in WinForms for example : 

```
if (InvokeRequired)
                BeginInvoke(new EventHandler(Instance_CurrentModeChanged), sender, e);
            else
                btnNextPile.Enabled = !ClientControl.Instance.IsHammerStarted;
```

In Qt, we can get the QThread in which the object belongs to. In C# there's no such things and that's sad :( the async stuff are not suited for my job and I don't like them.

**Reply by HaloFour** [reply]
Sounds like [`SynchronizationContext`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.synchronizationcontext?view=net-5.0), which is more a general-purpose thread marshaling API akin to `BeginInvoke` from WinForms.  It's also what TPL uses under the hood when `await`ing a `Task<T>`.  You're correct that C# is totally agnostic to thread affinity.

**Reply by embeddedmz** [reply]
I have already read these articles related to SynchronizationContext : https://devblogs.microsoft.com/pfxteam/await-synchronizationcontext-and-console-apps/ and https://devblogs.microsoft.com/pfxteam/await-synchronizationcontext-and-console-apps-part-2/ but I didn't succeed in using them in a simple example (a thread sending to another thread lambdas to be executed). Finally, I found that the closest thing to what it can be done in the Qt framework, is using a blocking collection of Action. Under the hood, that's how queued signals are made in Qt.

**Reply by theunrepentantgeek** [reply]
In the C# world, `SynchronizationContext` is definitely the way to do this. I remember doing this at at previous employer and recall it was a bit fiddly to set up but worked perfectly once I understood the details.

I found [this](https://stackoverflow.com/questions/39271492/how-do-i-create-a-custom-synchronizationcontext-so-that-all-continuations-can-be) question on StackOverflow that seems directly relevant to your needs. The answer is by Stephen Cleary - his [blog](https://blog.stephencleary.com/) is a goldmine of information about concurrency and async in .NET.

The chances of a new language feature to provide a different way to achieve something that's _already well supported_ are extremely slim (even more so given the requirement is a rare one).

**Reply by TonyValenti** [reply]
Instead of a blocking collection, you might look at system.threading.channels.

**Comment by embeddedmz** [comment]
Thank you all for all these useful informations. With my colleagues, we will look at the informations you shared with me (channels look promising) and try to see what can be useful for our use cases (serialize requests in a thread, etc...)

**Comment by embeddedmz** [comment]
Is there a .NET documentation like this : https://doc.qt.io/qt-5/threads-technologies.html to advise C# developers o, the available threading techniques within C# ?

**Comment by embeddedmz** [comment]
I made this example that may be of interest to some https://github.com/embeddedmz/message_passing_on_csharp
